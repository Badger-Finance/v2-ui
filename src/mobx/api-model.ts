/**
 * All API Model Objects should be directly in line with API interface model
 * objects to ensure data compatiblity. Interfaces located at link below:
 * https://github.com/axejintao/badger-api/tree/staging/src/interface
 */
import BigNumber from "bignumber.js";

export interface Geyser {
  emissions: Emission[],
}

export interface Token {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
}

export interface Emission {
  token: Token,
  unlockSchedule: UnlockSchedule,
}

export interface UnlockSchedule {
	initialLocked: BigNumber,
	endAtSec: BigNumber,
	durationSec: BigNumber,
	startTime: BigNumber,
}

export interface TokenBalance {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
  balance: number,
  value: number,
}

export interface ValueSource {
  name: string,
  apy: number,
  performance: Performance,
}

export interface Sett {
  name: string,
  asset: string,
  ppfs: number,
  value: number,
  apy: number,
  tokens: TokenBalance[],
  sources: ValueSource[],
  geyser?: Geyser,
}

export interface SettSummary {
  name: string,
  asset: string,
  value: number,
  tokens: TokenBalance[],
}
