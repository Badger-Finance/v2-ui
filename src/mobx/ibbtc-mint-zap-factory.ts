import { IbBTCMintZap } from './model/vaults/ibbtc-mint-zap';
import { RootStore } from './RootStore';
import addresses from 'config/ibBTC/addresses.json';
import { TokenZap } from './model/vaults/token-zap';
import { GeneralVaultZap } from './model/vaults/general-vault-zap';
import { RenVaultZap } from './model/vaults/ren-vault-zap';
import { Token } from '@badger-dao/sdk';

export class IbBTCMintZapFactory {
  private static tokenZaps: Record<string, TokenZap> = {};
  private static generalVaultZaps: Record<string, GeneralVaultZap> = {};
  private static renVaultZaps: Record<string, RenVaultZap> = {};

  static getIbBTCZap(store: RootStore, token: Token): IbBTCMintZap {
    if (addresses.mainnet.contracts.TokenZap.supportedTokens.includes(token.address)) {
      if (!this.tokenZaps[token.address]) {
        this.tokenZaps[token.address] = new TokenZap(store, token);
      }
      return this.tokenZaps[token.address];
    }

    if (addresses.mainnet.contracts.GeneralVaultZap.supportedTokens.includes(token.address)) {
      if (!this.generalVaultZaps[token.address]) {
        this.generalVaultZaps[token.address] = new GeneralVaultZap(store, token);
      }
      return this.generalVaultZaps[token.address];
    }

    if (addresses.mainnet.contracts.RenVaultZap.supportedTokens.includes(token.address)) {
      if (!this.renVaultZaps[token.address]) {
        this.renVaultZaps[token.address] = new RenVaultZap(store, token);
      }
      return this.renVaultZaps[token.address];
    }

    throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
  }
}
