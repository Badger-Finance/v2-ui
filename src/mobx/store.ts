import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
