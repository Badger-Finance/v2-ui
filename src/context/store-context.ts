import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;