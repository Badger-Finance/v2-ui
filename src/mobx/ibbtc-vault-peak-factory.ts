import { IbbtcVaultPeak } from './model/vaults/ibbtc-vault-peak';
import { RootStore } from './store';
import addresses from 'config/ibBTC/addresses.json';
import { ZapPeak } from './model/vaults/zap-peak';
import { YearnPeak } from './model/vaults/yearn-peak';
import { BadgerPeak } from './model/vaults/badger-peak';
import { IbbtcOptionToken } from './model/tokens/ibbtc-option-token';

export class IbbtcVaultPeakFactory {
	static createIbbtcVaultPeakForToken(store: RootStore, token: IbbtcOptionToken): IbbtcVaultPeak {
		if (addresses.mainnet.contracts.ZapPeak.supportedTokens.includes(token.symbol)) {
			return new ZapPeak(store, token);
		}

		if (addresses.mainnet.contracts.yearnWBTCPeak.supportedTokens.includes(token.symbol)) {
			return new YearnPeak(store, token);
		}

		if (addresses.mainnet.contracts.BadgerSettPeak.supportedTokens.includes(token.symbol)) {
			return new BadgerPeak(store, token);
		}

		throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
	}
}
