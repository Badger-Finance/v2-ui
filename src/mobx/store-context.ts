import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
