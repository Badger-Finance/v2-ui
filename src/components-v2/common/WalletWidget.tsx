import { Button, makeStyles } from '@material-ui/core';
import useENS from 'hooks/useEns';
import { StoreContext } from 'mobx/stores/store-context';
import { observer } from 'mobx-react-lite';
import React, { useContext } from 'react';

import { shortenAddress } from '../../utils/componentHelpers';

const useStyles = makeStyles((theme) => ({
	walletDot: {
		display: 'block',
		width: theme.spacing(0.9),
		height: theme.spacing(0.8),
		marginLeft: theme.spacing(0.4),
		borderRadius: theme.spacing(0.4),
	},
	redDot: {
		background: theme.palette.error.main,
	},
	greenDot: {
		background: theme.palette.success.main,
	},
	walletButtonLabel: {
		textTransform: 'none',
	},
}));

const WalletWidget = observer(() => {
	const classes = useStyles();
	const store = useContext(StoreContext);
	const { uiState, wallet } = store;

	console.log(wallet.address);

	async function connect(): Promise<void> {
		if (wallet.isConnected) {
			uiState.toggleWalletDrawer();
		} else {
			try {
				await wallet.connect();
			} catch (error) {
				const isModalClosed = String(error).includes('User closed modal');
				const isEmptyAccounts = String(error).includes('Error: accounts received is empty');
				if (!isModalClosed && !isEmptyAccounts) {
					console.error(error);
					uiState.queueError('Issue connecting, please try again');
				}
			}
		}
	}

	const { ensName } = useENS(wallet.address);
	const walletAddress = wallet.address ? shortenAddress(wallet.address) : 'Connect';

	return (
		<Button
			disableElevation
			color="primary"
			variant={wallet.isConnected ? 'outlined' : 'contained'}
			onClick={connect}
			classes={{ label: classes.walletButtonLabel }}
		>
			{ensName || walletAddress}
		</Button>
	);
});

export default WalletWidget;
