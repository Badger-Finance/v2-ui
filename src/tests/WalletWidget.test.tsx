/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { cleanup, customRender } from './Utils';
import WalletWidget from '../components-v2/common/WalletWidget';
import '@testing-library/jest-dom';
import { StoreProvider } from '../mobx/store-context';
import store from '../mobx/RootStore';
import { WalletStore } from '../mobx/stores/WalletStore';

describe('WalletWidget', () => {
	afterEach(cleanup);

	const testStore = store;

	test('Renders correctly', () => {
		const { container } = customRender(
			<StoreProvider value={testStore}>
				<WalletWidget />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	test('Renders mobile version correctly', () => {
		function createMatchMedia() {
			return (): MediaQueryList => ({
				media: '480px',
				matches: true,
				addEventListener: () => {},
				addListener: () => {},
				removeListener: () => {},
				removeEventListener: () => {},
				onchange: () => {},
				dispatchEvent: () => true,
			});
		}

		window.matchMedia = createMatchMedia();

		const { container } = customRender(
			<StoreProvider value={testStore}>
				<WalletWidget />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	//TODO: re-add test after WalletWidget is refactored
	// test('Displays walletSelect menu upon click', async () => {
	// 	customRender(
	// 		<StoreProvider value={testStore}>
	// 			<WalletWidget />
	// 		</StoreProvider>,
	// 	);
	// 	fireEvent.click(screen.getByText('Connect'));
	// 	// Checks that menu openned by finding the MetaMask option
	// 	expect(await screen.findByText('MetaMask')).toMatchSnapshot();
	// });

	test('Connected address is properly displayed', async () => {
		jest.spyOn(WalletStore.prototype, 'isConnected', 'get').mockReturnValue(true);
		jest.spyOn(WalletStore.prototype, 'address', 'get').mockReturnValue(
			'0x1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a',
		);
		const { container } = customRender(
			<StoreProvider value={testStore}>
				<WalletWidget />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});
});
