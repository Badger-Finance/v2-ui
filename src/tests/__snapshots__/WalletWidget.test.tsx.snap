// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletWidget Connected address is properly displayed 1`] = `
<div>
  <button
    class="MuiButtonBase-root-35 MuiButton-root-6 MuiButton-text-8 makeStyles-walletButton-4 MuiButton-textPrimary-9 MuiButton-disableElevation-17"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-7 makeStyles-walletButtonLabel-5"
    >
      0x1a..1a1a
    </span>
    <span
      class="MuiTouchRipple-root-38"
    />
  </button>
</div>
`;

exports[`WalletWidget Displays walletSelect menu upon click 1`] = `
<span
  class="svelte-1799bj2"
>
  MetaMask
   
</span>
`;

exports[`WalletWidget Renders correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root-35 MuiButton-root-6 MuiButton-outlined-11 makeStyles-walletButton-4 MuiButton-disableElevation-17"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-7 makeStyles-walletButtonLabel-5"
    >
      Connect
    </span>
    <span
      class="MuiTouchRipple-root-38"
    />
  </button>
</div>
`;

exports[`WalletWidget Renders mobile version correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root-35 MuiButton-root-6 MuiButton-text-8 makeStyles-walletButton-4 MuiButton-textPrimary-9 MuiButton-disableElevation-17"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-7 makeStyles-walletButtonLabel-5"
    >
      Connect
    </span>
    <span
      class="MuiTouchRipple-root-38"
    />
  </button>
</div>
`;
